src/emu/sound.cpp

```diff
@@ -42,288 +44,544 @@ const attotime sound_manager::STREAMS_UPDATE_ATTOTIME = attotime::from_hz(STREAM
+
+void sound_stream_input::apply_sample_rate_changes(u32 updatenum, u32 downstream_rate)
+{
+	// shouldn't get here unless valid
+	sound_assert(valid());
+
+	// if we have a resampler, tell it (and it will tell the native source)
+	if (m_resampler_source != nullptr)
+		m_resampler_source->stream().apply_sample_rate_changes(updatenum, downstream_rate);
+
+	// otherwise, just tell the native source directly
+	else
+		m_native_source->stream().apply_sample_rate_changes(updatenum, downstream_rate);
 }
 
 
-void sound_stream::sync_update(void *, s32)
+
+//**************************************************************************
+//  SOUND STREAM
+//**************************************************************************
+
+//-------------------------------------------------
+//  sound_stream - constructor with old-style
+//  callback
+//-------------------------------------------------
+
+sound_stream::sound_stream(device_t &device, u32 inputs, u32 outputs, u32 output_base, u32 sample_rate, stream_update_legacy_delegate callback, sound_stream_flags flags) :
+	m_device(device),
+	m_next(nullptr),
+	m_sample_rate((sample_rate < SAMPLE_RATE_OUTPUT_ADAPTIVE) ? sample_rate : 48000),
+	m_pending_sample_rate(SAMPLE_RATE_INVALID),
+	m_last_sample_rate_update(0),
+	m_input_adaptive(sample_rate == SAMPLE_RATE_INPUT_ADAPTIVE),
+	m_output_adaptive(sample_rate == SAMPLE_RATE_OUTPUT_ADAPTIVE),
+	m_synchronous((flags & STREAM_SYNCHRONOUS) != 0),
+	m_sync_timer(nullptr),
+	m_input(inputs),
+	m_input_array(inputs),
+	m_input_view(inputs),
+	m_empty_buffer(100),
+	m_output_base(output_base),
+	m_output(outputs),
+	m_output_array(outputs),
+	m_output_view(outputs),
+	m_callback(std::move(callback)),
+	m_callback_ex(stream_update_delegate(&sound_stream::oldstyle_callback_ex, this))
 {
-	update();
-	attotime time = m_device.machine().time();
-	attoseconds_t next_edge = m_attoseconds_per_sample - (time.attoseconds() % m_attoseconds_per_sample);
-	m_sync_timer->adjust(attotime(0, next_edge));
+	// common init
+	init_common(inputs, outputs, sample_rate, flags);
+}
+
+
+//-------------------------------------------------
+//  sound_stream - constructor with new-style
+//  callback
+//-------------------------------------------------
+
+sound_stream::sound_stream(device_t &device, u32 inputs, u32 outputs, u32 output_base, u32 sample_rate, stream_update_delegate callback, sound_stream_flags flags) :
+	m_device(device),
+	m_next(nullptr),
+	m_sample_rate((sample_rate < SAMPLE_RATE_OUTPUT_ADAPTIVE) ? sample_rate : 48000),
+	m_pending_sample_rate(SAMPLE_RATE_INVALID),
+	m_input_adaptive(sample_rate == SAMPLE_RATE_INPUT_ADAPTIVE),
+	m_output_adaptive(sample_rate == SAMPLE_RATE_OUTPUT_ADAPTIVE),
+	m_synchronous((flags & STREAM_SYNCHRONOUS) != 0),
+	m_sync_timer(nullptr),
+	m_input(inputs),
+	m_input_array(inputs),
+	m_input_view(inputs),
+	m_output_base(output_base),
+	m_output(outputs),
+	m_output_array(outputs),
+	m_output_view(outputs),
+	m_callback_ex(std::move(callback))
```

Comment:
There’s a lot of repetition here – is it not possible to have a protected or private constructor that does most of the work and delegate to it from the public constructors?

Reply:
Probably. Yeah that kind of bugged me as well. Good suggestion.